packages/wasm-utils/src/note/secrets.rs

```diff
@@ -0,0 +1,49 @@
+use ark_std::rand::rngs::OsRng;
+use arkworks_circuits::setup::mixer::{setup_leaf, setup_leaf_with_privates_raw};
+use arkworks_gadgets::prelude::ark_bls12_381::Fq as BlsFr;
+use arkworks_gadgets::prelude::ark_bn254::Fq as Bn254Fr;
+use arkworks_gadgets::prelude::*;
+use arkworks_utils::utils::common::Curve as ArkworksCurve;
+
+use crate::types::{Curve, OpStatusCode};
+use crate::utils::get_hash_params_x5;
+
+pub fn generate_secrets(exponentiation: usize, curve: Curve, rng: &mut OsRng) -> Result<Vec<u8>, OpStatusCode> {
+	let secrets: Vec<u8> = match (curve, exponentiation) {
+		(Curve::Bls381, 5) => {
+			let (params_5, ..) = get_hash_params_x5::<BlsFr>(ArkworksCurve::Bls381);
+			let (secret_bytes, nullifier_bytes, ..) = setup_leaf(&params_5, rng);
+			[secret_bytes, nullifier_bytes].concat()
+		}
+		(Curve::Bn254, 5) => {
+			let (params_5, ..) = get_hash_params_x5::<Bn254Fr>(ArkworksCurve::Bn254);
+			let (secret_bytes, nullifier_bytes, ..) = setup_leaf(&params_5, rng);
+			[secret_bytes, nullifier_bytes].concat()
+		}
+		(..) => {
+			unreachable!("unreachable exponentiation {}", exponentiation)
+		}
+	};
+	Ok(secrets)
+}
+pub fn get_leaf_with_private_raw(curve: Curve, _exponentiation: i8, raw: &[u8]) -> (Vec<u8>, Vec<u8>) {
+	let secrets = raw[..32].to_vec();
```

Comment:
We should also ensure that `raw.len() >= 64`

Reply:
Added this